src/fluree/db/query/fql.cljc

```diff
@@ -618,16 +618,21 @@
                           res)]
                (if inVector? [res'] res'))
 
-             (let [res+agg (if aggregates (analytical/add-aggregate-cols res aggregates) res)
-                   offset  (if groupBy 0 offset)
-                   {:keys [headers tuples]} (if orderBy
-                                              (order-offset-and-limit-results orderBy res+agg offset group-limit)
-                                              res+agg)
-                   res     (<? (format-filter-tuples db tuples select-spec headers vars (dissoc opts :limit :offset :orderBy :groupBy)))]
+             (let [{:keys [headers tuples]} (if aggregates (analytical/add-aggregate-cols res aggregates) res)
+                   offset' (if (or selectDistinct? groupBy) 0 offset)
+                   limit'  (if selectDistinct? nil group-limit)
+                   tuples' (if orderBy
+                             (order-offset-and-limit-results tuples headers orderBy offset' limit')
+                             tuples)
+                   res'    (->> (dissoc opts :limit :offset :orderBy :groupBy)
+                                (format-filter-tuples db tuples' select-spec headers vars)
+                               <?)]
                ;; TODO - drop unused columns, and calculate distinct before resolving all vals
                (if selectDistinct?
-                 (->> (into #{} res) (into []))
-                 res))))))
+                 (cond->> (distinct res')
+                          (and offset (< 0 offset)) (drop offset)
+                          (and group-limit (< 0 group-limit)) (take group-limit))
+                 res'))))))
```

Comment:
If I'm reading this right, you could end up applying the offset and limit twice if you have both orderBy and selectDistinct? Is that a known issue that you're saving for later refactoring, or am I reading it wrong?

Reply:
Above that section of code, where we are applying the sort to the results, we actually clear out the values:

                   offset' (if (or selectDistinct? groupBy) 0 offset)
                   limit'  (if selectDistinct? nil group-limit)
                   tuples' (if orderBy
                             (order-offset-and-limit-results tuples headers orderBy offset' limit')
                             tuples)